### PR TITLE
[Frontend][Misc] Don't Repeat Yourself (DRY) Sampling

### LINKED ISSUE #8581: [Feature]: DRY Sampling
### 🚀 The feature, motivation and pitch

DRY is a sampler that completely mitigates repetitions. This is especially important for small models which tend to slop in large contexts. Here's an explanation of DRY from the author himself https://github.com/oobabooga/text-generation-webui/pull/5677
Along with oobabooga, koboldcpp also has an implementation of DRY which according to author IIRC is better than oobabooga's
DRY has been a completely game changer for me and from what I have seen several others. It completely removes need for other samplers like top_p, top_k, repetition_penalty. It is recommended to be used with min_p and produces great coherent results.

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by AlpinDale**
FIX #8581

This PR adds support for the DRY sampler, a modern repetition penalty to prevent repetitive outputs at a sequence-level, and performs much better than older penalty methods.

## Showcase:

Prompt:
```
All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work
```

### No DRY output:
```
 and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy. All work and no play makes Jack a dull boy.\n\nSomewhere between the repetition and the mundane task of George Orwell’s 1984 and the evocative yet disturbing image of a mess of typewriter keys
```

### dry_multiplier=1.0
```
... (more)\nIn a remote part of town there once lived a lonely young boy named Danny who wrote the phrase \"All work and no p(l)ay makes Jack a d(u)ll boy.\" on a typewriter for what he would shortly come to realize was the remainder of his life. Danny, who was 8 years old at the time (1984), had not yet learned how to turn off his own typed phrase.\nAs I, a good six years older (1990), like many others who experienced The Shining, witnessed the two year and two month documented torment that befell this isolated young soul. A remote"
```

Currently, using DRY can incur an overhead of up to ~25%. We can probably optimize this.

### CODE REVIEW

File: vllm/model_executor/layers/sampler.py [hunk 1]
```diff
@@ -384,6 +395,116 @@ def _apply_min_tokens_penalty(
...
+    multipliers: torch.Tensor,
+    bases: torch.Tensor,
+    allowed_lengths: torch.Tensor,
+    sequence_breakers_ids: torch.Tensor,
+    ranges: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Apply Don't Repeat Yourself (DRY) sampling to the logits.
+
+    Reference: https://github.com/PygmalionAI/aphrodite-engine/blob/a3c03db7355b33c0dfd670b084e827d0aa7442d1/aphrodite/modeling/layers/sampler.py#L621-L702
```
Comment (p-e-w) [inline vllm/model_executor/layers/sampler.py:434]: Probably makes more sense to link to the original PR (https://github.com/oobabooga/text-generation-webui/pull/5677), as that contains a detailed description and discussion, whereas this is just code similar to the code below.

File: vllm/model_executor/layers/sampler.py [hunk 2]
```diff
@@ -384,6 +395,116 @@ def _apply_min_tokens_penalty(
...
+    """
+
+    VOCAB_SIZE = logits.size(-1)
+    MAX_NGRAM = 100
```
Comment (p-e-w) [inline vllm/model_executor/layers/sampler.py:438]: 50 is more than enough. In practice, top logits tend to be on the order of 10^2, and 1.75^50 is already 10^12. This will double performance for pathological inputs.

File: vllm/model_executor/layers/sampler.py [hunk 3]
```diff
@@ -384,6 +395,116 @@ def _apply_min_tokens_penalty(
...
+
+        # If [token] is picked, what's the longest ngram that would match?
+        ngram_lens = torch.zeros(VOCAB_SIZE,
+                                 dtype=torch.int32,
+                                 device=logits.device)
+
+        # Find all instances of the last token- potential ngrams!
+        endpoint_indexes = torch.nonzero(token_seq == last_token,
+                                         as_tuple=True)[0].tolist()
+        # NOTE(alpin): This seems like the slow part.
```
Comment (p-e-w) [inline vllm/model_executor/layers/sampler.py:485]: Here's an idea: Instead of matching a single token above with `nonzero`, try to match the last `allowed_length` tokens. Shorter matches don't need to be penalized by definition, and this should drastically cut down on the number of matches that must be looped over. I'm not sure how to perform subsequence matching in PyTorch, but there should be a way to do this as a vectorized operation.

File: vllm/model_executor/layers/sampler.py [hunk 4]
```diff
@@ -384,6 +395,116 @@ def _apply_min_tokens_penalty(
...
+                    break
+                if token_seq[idx - unwind] != token_seq[-unwind - 1]:
+                    break
+            next_tok = token_seq[idx + 1]
+            # The repeated tokens BEFORE next_tok (+1 to include [idx]).
+            ngram_lens[next_tok] = max(ngram_lens[next_tok].item(), unwind + 1)
+
+        # Convert ngram lengths to penalty exponents
+        penalty_mask = ngram_lens > 0
+        scales = bases[irow]**(ngram_lens[penalty_mask] - min_ngram)
```
Comment (p-e-w) [inline vllm/model_executor/layers/sampler.py:500]: What if `ngram_lens[penalty_mask] - min_ngram < 0`? Perhaps I'm misunderstanding the implementation, but that case doesn't appear to be accounted for.

Comment (akkikiki) [reply vllm/model_executor/layers/sampler.py:500]: +1 on confirming this.
If I understand correctly, we want something like the following to filer out the cases when `ngram_lens[penalty_mask] - min_ngram < 0`:
```
# Apply penalties
penalty_mask = ngram_lens > 0
exponent = ngram_lens[penalty_mask] - min_ngram

# Ensure that when exponent < 0, scales is set to 0
valid_exponent_mask = exponent >= 0
scales = torch.zeros_like(exponent, dtype=torch.float32)
scales[valid_exponent_mask] = bases[irow] ** exponent[valid_exponent_mask]

# Apply penalties
logits[irow][penalty_mask] -= multipliers[irow] * scales
```
